### PR TITLE
Cubeviz image viewer coordinates display

### LINKED ISSUE #1090: Cubeviz should have coordinates display like Imviz
**Reporter:** pllim

As a user, I want to see the information under my cursor when I mouseover the datacube display. Unlike Imviz, the plugin needs to be aware which 3D slice is currently shown (Imviz understand the concept of each 2D image as a glue data layer, but Cubeviz is a different beast).

Can probably copy a lot of stuff from the Imviz implementation, but not all of them. Actually, maybe might be able to use the same plugin but the viewer hook-up would be different.

[:cat:](https://jira.stsci.edu/browse/JDAT-1960) [👻](https://jira.stsci.edu/browse/JDAT-2151)

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to add Imviz-like coordinates display for Cubeviz image viewers.

This is all UI so I am not sure how to add tests. There is currently no unit test for this equivalent capability in Imviz. Reviewers should interactively load a few different datasets that we are supposed to support and see what happens...

Here is a screenshot with data from Cubeviz example notebook (the mouse was at the center of flux viewer for this one):

Screenshot 2022-05-12 132328

Fixes #1090 

Imviz's display was originally done by @astrofrog (e.g., #557).

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)? (N/A)
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by kecnry** [comment]
Any chance this can be generalized (without too much effort) and moved to a Mixin that both cubeviz and imviz viewers inherit from (especially if adding the same to specviz is in our future)?  It seems like a lot of the code is the same.

**Comment by pllim** [comment]
Re: Refactoring -- I'd rather wait to see what needs to happen for #1299 first.

**Comment by kecnry** [comment]
Ok, do you want to move ahead with reviews on this as-is (ignoring refactoring) then and deal with that as a separate effort?  Or should we put this PR on hold until 1299 is resolved?

**Comment by pllim** [comment]
> do you want to move ahead with reviews

I'd prefer we move ahead as-is. Thanks!

File: jdaviz/configs/cubeviz/plugins/viewers.py
Comment (kecnry) [inline jdaviz/configs/cubeviz/plugins/viewers.py:67]: how is `visible_layers` ordered?  If there are multiple layers and they're linked correctly, the coordinates shouldn't care what layer, but the _value_ will... is this guaranteed to be the "top" layer if multiple layers are plotted without changing transparency?  Is there any room to squeeze in the layer name for cases where `visible_layers` includes more than one?

This behavior/situation might be something that is worth adding test coverage, if possible.

**Review by kecnry — COMMENTED** [review]
> I'd prefer we move ahead as-is.

Ok, then let's refactor if/when we add to any other viz or when 1299 is resolved.  I don't love the hit to coverage, but the interactive nature is difficult to test without more UI testing infrastructure.  At least this is something that will be very obvious if it completely breaks (although very non-obvious if it just displays the wrong information).  Is it possible to fake the `data` sent to `on_mouse_or_key_event` to at least make sure the value and a single known pixel is parsed correctly?

File: jdaviz/configs/cubeviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/viewers.py:67]: The inner workings of visible layers, I'd have to defer to @astrofrog or @maartenbreddels. I wish there is an easier way for me to reliably grab this info too but then Glue lets you overlay this and that, which complicates things. This particular info panel would ignore the semi-transparent overlay use cases.

Re: tests -- the info panel is kinda tested non-interactively in https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/imviz/tests/test_linking.py (the `label_mouseover` stuff). Not that the tests are exhaustive because they obviously missed #1299 but again adding a test for that would be out of scope here. Still, I guess I can add similar test(s) to Cubeviz...

Comment (kecnry) [reply jdaviz/configs/cubeviz/plugins/viewers.py:67]: from trial and error, it seems that this does always show the "top" layer, which is the most sensible (simple) behavior.  If there is any room, I do think a label would be useful, but will approve this and that can always be added later if you don't have time to include it in scope here (as I think the same thing could apply to the existing display in imviz).

**Review by kecnry — APPROVED** [review]

File: jdaviz/configs/cubeviz/plugins/viewers.py
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/viewers.py:67]: Re: label -- In Imviz, I incorporated that into Compass plugin (because the label just seems to go with Compass naturally). Originally when I was new and naive, I attempted glue-viz/glue-jupyter#244 but failed. I don't think there is any more space for label in the coordinates display, as it is already pretty crowded.

This does bring up a good point... What if someone deselects the cube and have a 2D moment map displayed in the cube viewer instead? Is that a use case we even support?

File: jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:12]: Since I am updating this test anyway, I took the liberty to make it more realistic in loading like a user would.

File: jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py:31]: But in changing from using `app.add_data` to `cubeviz_helper.load_data`, the moment map shape changed! @rosteen , since you refactored how Cubeviz swaps dimensions around, does this make sense to you?

File: jdaviz/configs/cubeviz/plugins/parsers.py
Comment (pllim) [inline jdaviz/configs/cubeviz/plugins/parsers.py:78]: It is highly annoying to me that `data_label` is ignored when you pass `Spectrum1D` into Cubeviz, so I fixed it here. Unless there is a good reason not to do this? I cannot think of any good reason.

**Review by pllim — COMMENTED** [review]
@kecnry , I added tests but might also have opened a small rabbit hole. Please re-review. Thanks!

**Pull request merged**
